### ISSUE TITLE
Charging/Billing Plugin for RestComm

### ISSUE THREAD

**Issue by deruelle** [RestComm/Restcomm-Connect#159]
Currently there is no way to charge accounts for a given call or SMS or app installation from RAS.
We can implement some API and plugin at RestComm level to interconnect with different billing systems the same way it is done for phone number provisioning.

**Comment by deruelle** [comment]
We can provide a few plugins out of the box such as H20, Meveo (for pure FOSS play) and to make sure the plugin SPI is correctly defined.

**Comment by ivelin** [comment]
Agree. The main questions is which ones to prioritize first? H20 is US focused. ElephantTalk claims strength in International Markets.

Is Meveo the leading FOSS option? 
http://www.cio.com.au/article/324595/5_open_source_billing_systems_watch/

**Comment by ivelin** [comment]
http://www.manaty.net/products/meveo-open-source-billing-system

**Comment by deruelle** [comment]
the website is http://meveo.info/

**Comment by ivelin** [comment]
Sounds like a good start. We would have to see what scenarios make most sense to users of both systems.

The first two use cases make sense to me:
1. When an account is created in Meveo, trigger provisioning of an account in Restcomm. 
2. When an access point is created in Meveo, provision a matching Phone Number in Restcomm.

As far as phone call and SMS event notifications, it may be overwhelming to issue external web service requests each time a CDR record is logged. We may start with a more traditional approach of Meveo sweeping CDR logs once an hour or so to generate the respective billing records in Meveo.

Restcomm has a REST API for retrieving call records, but it will be limiting to use that for billing at scale.

A next step could be integrating Restcomm and Meveo via Diameter, which is designed for this purpose. Restcomm can fire a Diameter datagram when a CDR is logged. Eventually we can look at pre-paid scenarios. Those can be driven by further market demand.

**Comment by manaty** [comment]
Yes it's a good plan.
We will add Restcomm provisionning and mediation to the feature list of Meveo v3.3 (available next month), prepaid and realtime charging is schedule in 4.0 for mid-2015
Can Restcomm produce the CDR logs in our format : dateTime; quantity; accessId; param1...;param4 ?
Could you provide a demo env that we could plug to Meveo demo env (http://demo.meveo.info) ?

**Comment by deruelle** [comment]
RestComm produces CDRs for Calls that have this format http://docs.telestax.com/restcomm-api-calls/ and for SMS http://docs.telestax.com/restcomm-api-sms-messages/. 
@gvagenas might be able to point you to the DB schema structure.

On the demo env, you can either download RestComm itself from https://mobicents.ci.cloudbees.com/job/RestComm/lastSuccessfulBuild/artifact/ or use the AMI on https://aws.amazon.com/marketplace/pp/B00FFHJ6SU, would that be enough for the integration ?

**Comment by manaty** [comment]
We succesfully integrated Meveo with Restcomm for provisionning of Accounts and PhoneNumbers (both in creation and update mode) using webhooks of Meveo 4.0 version.
We sill release a short tutorial for that part.
Now about CDRs, it is not convenient that meveo calls an API to retrieve them, it would be better that Restcom either produce cdr files that can be transfered to meveo (for postpaid) , or call meveo realtime charging services (for prepaid) : any suggestion for that part ?

**Comment by manaty** [comment]
you can find the video here : http://youtu.be/8MQYIWm04vc

**Comment by ivelin** [comment]
Excellent work! Broadcasted on social media.

Would you please create a follow up screencast that showcases the billing strengths of Meveo:
- Show how a call going through Restcomm is charged to a user in Meveo
- Pre-paid vs post-paid
- Invoice generation
- Agent commission allocation
- Pricing modes - tiered, graded, etc.
- Reporting, analytics, dashboard

If there are screencast for that already out there, please forgive my ignorance. Just point me to the links.

Finally, would you be interested in writing a guest blog post about it? If so, please send me your email and I will generate registration invitation to the Telestax blog.

Thank you,

Ivelin

**Comment by ivelin** [comment]
@manaty, just noticed your questions about CDRs for post-paid and Restcomm authorization requests for pre-paid.

Do you implement Diameter? We can integrate Restcomm with Meveo via the Diameter Charging Server or directly via jDiameter if that makes more sense. The Ro (pre-paid) and Rf (post-paid) interfaces are designed for the purpose. More efficient than moving around large chunks of CRD files: 
https://code.google.com/p/charging-server/

@ammendonca is the Diameter expert. He needs to chime in here.

**Comment by manaty** [comment]
Hi Ivelin

We don't currently implement diameter, we have a REST API for realtime charging.
Might be not so difficult to implement your diameter stack if some people can help with testing.

Currently Restcomm do not create CDR files that we could transfer to Meveo ? that would allow me to complete the postpaid use case.

Best regards,
Sebastien

**Comment by manaty** [comment]
Ivelin,

Yes i'll do the screencast as soon as i can get Restcomm CDRs, and i'm also interested in writing a guest blog post.
please just use my manaty.net email

Best regards,
Sebastien

**Comment by vikasbo** [comment]
@deruelle We can meet the requirements against a customized version of [jBilling CE 4x](http://webdataconsulting.github.io/billing/). One can already process cdrs over rest/soap api on a customized version.

**Comment by deruelle** [comment]
@vikasbo very interesting. Can you sign the Contributor License Agreement at https://telestax.com/open-source/#Contribute and do a pull request for that work ?

Did you implement that as an external batch to process the CDRs in a post paid fashion or you implemented a real time engine with Diameter Ro for pre paid ?

**Comment by netg5** [comment]
Dear Senior Dev
So would you please give what would be better integration for 1 million /sec prepaid transaction handler is that jDiameter / API will be perfect suite for Telco ?
